File: src/Python/devpossible_ton/devpossible_ton/parser/ton_parser.py
"""
TonParser - Parser for TON format
Copyright (c) 2024 DevPossible, LLC
"""

from dataclasses import dataclass
from typing import Optional, Any
from ..lexer import TonLexer, Token, TokenType
from ..models import TonDocument, TonObject, TonValue, TonArray
from ..errors import TonParseError


@dataclass
class TonParseOptions:
    """Options for parsing TON format."""
    allow_trailing_comma: bool = False
    strict_mode: bool = True
    allow_partial: bool = False


class TonParser:
    """Parser for TON format."""

    def __init__(self, options: Optional[TonParseOptions] = None, **kwargs):
        # Handle keyword arguments for backward compatibility
        if options is None and kwargs:
            options = TonParseOptions(**kwargs)
        self.options = options or TonParseOptions()
        self.tokens = []
        self.current = 0

    def parse(self, text: str) -> TonDocument:
        """Parse TON text into a document."""
        lexer = TonLexer(text)
        self.tokens = lexer.tokenize()
        self.current = 0

        root = self._parse_value()

        if not self._is_at_end():
            token = self._peek()
            raise TonParseError(
                'Unexpected content after parsing',
                token.line,
                token.column
            )

        return TonDocument(root)
    
    def parse_file(self, file_path: str) -> TonDocument:
        """Parse a TON file into a document."""
        if not file_path:
            raise ValueError("file_path cannot be None or empty")
        
        import os
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        return self.parse(content)
    
    async def parse_file_async(self, file_path: str) -> TonDocument:
        """Parse a TON file into a document asynchronously."""
        import aiofiles
        import os
        
        if not file_path:
            raise ValueError("file_path cannot be None or empty")
        
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        
        async with aiofiles.open(file_path, 'r', encoding='utf-8') as f:
            content = await f.read()
        
        return self.parse(content)

    def _parse_value(self) -> Any:
        """Parse a value."""
        token = self._peek()

        if token.type == TokenType.LEFT_BRACE:
            return self._parse_object()
        elif token.type == TokenType.LEFT_BRACKET:
            return self._parse_array()
        elif token.type in (TokenType.STRING, TokenType.NUMBER, TokenType.BOOLEAN,
                            TokenType.NULL, TokenType.UNDEFINED, TokenType.GUID):
            return TonValue(self._advance().value)
        elif token.type in (TokenType.ENUM, TokenType.ENUM_SET):
            return self._parse_enum()
        elif token.type in (TokenType.STRING_HINT, TokenType.NUMBER_HINT,
                            TokenType.BOOLEAN_HINT, TokenType.DATE_HINT):
            return self._parse_hinted_value()
        elif token.type == TokenType.CLASS_NAME:
            return self._parse_typed_object()
        elif token.type == TokenType.END_OF_FILE:
            raise TonParseError(
                'Unexpected end of input',
                token.line,
                token.column
            )
        else:
            raise TonParseError(
                f'Unexpected token: {token.type}',
                token.line,
                token.column
            )

    def _parse_object(self) -> TonObject:
        """Parse an object."""
        self._consume(TokenType.LEFT_BRACE, 'Expected {')
        obj = TonObject()

        while not self._check(TokenType.RIGHT_BRACE) and not self._is_at_end():
            # Parse property name
            name_token = self._advance()
            if name_token.type not in (TokenType.IDENTIFIER, TokenType.STRING):
                raise TonParseError(
                    'Expected property name',
                    name_token.line,
                    name_token.column
                )

            name = name_token.value

            # Check for type annotation (name:type: value syntax)
            type_hint = None
            if self._check(TokenType.COLON):
                self._advance()  # consume first :
                if self._check(TokenType.IDENTIFIER):
                    type_hint = self._advance().value
                    # Consume the second colon if present
                    if self._check(TokenType.COLON):
                        self._advance()  # consume second :

            # Parse property value
            value = self._parse_value()

            if type_hint and isinstance(value, TonValue):
                value.type_hint = type_hint

            obj.set(name, value)

            # Check for comma or closing brace (commas are optional in pretty format)
            if not self._check(TokenType.RIGHT_BRACE):
                if self._check(TokenType.COMMA):
                    self._advance()
                # Commas are always optional - TON format allows newline-separated properties

        self._consume(TokenType.RIGHT_BRACE, 'Expected }')
        return obj

    def _parse_array(self) -> TonArray:
        """Parse an array."""
        self._consume(TokenType.LEFT_BRACKET, 'Expected [')
        arr = TonArray()

        while not self._check(TokenType.RIGHT_BRACKET) and not self._is_at_end():
            arr.add(self._parse_value())

            if not self._check(TokenType.RIGHT_BRACKET):
                if self._check(TokenType.COMMA):
                    self._advance()
                elif not self.options.allow_trailing_comma:
                    next_token = self._peek()
                    if next_token.type != TokenType.RIGHT_BRACKET:
                        raise TonParseError(
                            'Expected comma or ]',
                            next_token.line,
                            next_token.column
                        )

        self._consume(TokenType.RIGHT_BRACKET, 'Expected ]')
        return arr

    def _parse_enum(self) -> TonValue:
        """Parse an enum."""
        token = self._advance()
        return TonValue(token.value, 'enum')

    def _parse_hinted_value(self) -> TonValue:
        """Parse a hinted value."""
        hint_token = self._advance()
        type_hint_map = {
            TokenType.STRING_HINT: 'string',
            TokenType.NUMBER_HINT: 'number',
            TokenType.BOOLEAN_HINT: 'boolean',
            TokenType.DATE_HINT: 'date',
        }
        type_hint = type_hint_map.get(hint_token.type, 'unknown')

        value = self._parse_value()
        if isinstance(value, TonValue):
            value.type_hint = type_hint

        return value

    def _parse_typed_object(self) -> TonObject:
        """Parse a typed object."""
        class_name = self._advance().value

        # Check for instance count
        instance_count = None
        if self._check(TokenType.LEFT_PAREN):
            self._advance()  # consume (
            count_token = self._consume(TokenType.NUMBER, 'Expected instance count')
            instance_count = int(count_token.value)
            self._consume(TokenType.RIGHT_PAREN, 'Expected )')

        obj = self._parse_object()
        obj.class_name = class_name
        obj.instance_count = instance_count

        return obj

    def _consume(self, token_type: TokenType, message: str) -> Token:
        """Consume a token of the expected type."""
        if self._check(token_type):
            return self._advance()

        token = self._peek()
        raise TonParseError(message, token.line, token.column)

    def _check(self, token_type: TokenType) -> bool:
        """Check if current token is of expected type."""
        if self._is_at_end():
            return False
        return self._peek().type == token_type

    def _advance(self) -> Token:
        """Advance to next token."""
        if not self._is_at_end():
            self.current += 1
        return self._previous()

    def _peek(self) -> Token:
        """Peek at current token."""
        return self.tokens[self.current]

    def _previous(self) -> Token:
        """Get previous token."""
        return self.tokens[self.current - 1]

    def _is_at_end(self) -> bool:
        """Check if at end of tokens."""
        return self._peek().type == TokenType.END_OF_FILE

File: src/JavaScript/devpossible-ton/src/parser/TonParser.ts
/**
 * TonParser - Parser for TON format
 * Copyright (c) 2024 DevPossible, LLC
 */

import { TonLexer, Token, TokenType } from '../lexer/TonLexer';
import { TonDocument } from '../models/TonDocument';
import { TonObject } from '../models/TonObject';
import { TonValue } from '../models/TonValue';
import { TonArray } from '../models/TonArray';
import { TonParseOptions } from './TonParseOptions';
import { TonParseError } from '../errors/TonParseError';

// Re-export TonParseOptions for convenience
export { TonParseOptions };

export class TonParser {
  private tokens: Token[] = [];
  private current: number = 0;
  // @ts-ignore - options reserved for future use
  private options: TonParseOptions;

  constructor(options?: TonParseOptions) {
    this.options = options || {};
  }

  public parse(input: string): TonDocument {
    // Check for empty or whitespace-only input
    if (!input || input.trim().length === 0) {
      throw new TonParseError('Input cannot be empty', 1, 1);
    }

    const lexer = new TonLexer(input);
    this.tokens = lexer.tokenize();
    this.current = 0;

    // Parse optional header
    const header = this.parseHeader();

    const root = this.parseValue();

    if (!this.isAtEnd()) {
      throw new TonParseError(
        'Unexpected content after parsing',
        this.peek().line,
        this.peek().column
      );
    }

    const doc = new TonDocument(root);
    if (header) {
      doc.header = header;
    }
    return doc;
  }

  private parseHeader(): any {
    if (!this.check(TokenType.HeaderMarker)) {
      return null;
    }

    this.advance(); // consume #@

    const header: any = {};

    // Parse header properties (e.g., tonVersion = '1')
    while (!this.isAtEnd() && !this.check(TokenType.LeftBrace)) {
      // Check if we hit another marker or structural element
      if (this.check(TokenType.HeaderMarker)) {
        break;
      }

      // Parse property name
      const nameToken = this.advance();
      if (nameToken.type !== TokenType.Identifier &&
          nameToken.type !== TokenType.String &&
          nameToken.type !== TokenType.ClassName) {
        break; // Not a header property
      }
      const name = String(nameToken.value);

      // Expect = or :
      if (!this.check(TokenType.Equals) && !this.check(TokenType.Colon)) {
        break;
      }
      this.advance();

      // Parse value
      const value = this.parseHeaderValue();
      header[name] = value;

      // Optional comma
      if (this.check(TokenType.Comma)) {
        this.advance();
      }
    }

    return Object.keys(header).length > 0 ? header : null;
  }

  private parseHeaderValue(): any {
    const token = this.peek();

    if (token.type === TokenType.String) {
      return this.advance().value;
    } else if (token.type === TokenType.Number) {
      return this.advance().value;
    } else if (token.type === TokenType.Boolean) {
      return this.advance().value;
    } else if (token.type === TokenType.Identifier) {
      return this.advance().value;
    }

    return null;
  }

  private parseValue(): any {
    const token = this.peek();

    switch (token.type) {
      case TokenType.LeftBrace:
        return this.parseObject();

      case TokenType.LeftParen:
        // Check if this is a typed object (ClassName){...}
        return this.parseTypedObject();

      case TokenType.LeftBracket:
        return this.parseArray();

      case TokenType.String:
      case TokenType.Number:
      case TokenType.Boolean:
      case TokenType.Null:
      case TokenType.Undefined:
      case TokenType.Guid:
        return new TonValue(this.advance().value);

      case TokenType.Enum:
      case TokenType.EnumSet:
        return this.parseEnum();

      case TokenType.StringHint:
      case TokenType.NumberHint:
      case TokenType.BooleanHint:
      case TokenType.DateHint:
        return this.parseHintedValue();

      case TokenType.ClassName:
        return new TonValue(this.parseTypedObject());

      case TokenType.Identifier:
        // Check if this looks like a GUID pattern (contains hyphens)
        const identValue = token.value;
        if (typeof identValue === 'string' && identValue.includes('-')) {
          // Treat as a string value (e.g., "not-a-guid")
          this.advance();
          return new TonValue(identValue);
        }
        throw new TonParseError(
          'Expected { to start object',
          token.line,
          token.column
        );

      default:
        throw new TonParseError(
          `Unexpected token: ${token.type}`,
          token.line,
          token.column
        );
    }
  }

  private parseObject(): TonObject {
    this.consume(TokenType.LeftBrace, 'Expected {');

    // Check for typed object: {(ClassName)...}
    let className: string | undefined;
    let instanceCount: number | undefined;

    if (this.check(TokenType.LeftParen)) {
      this.advance(); // consume (

      if (!this.check(TokenType.ClassName) && !this.check(TokenType.Identifier)) {
        throw new TonParseError('Expected class name', this.peek().line, this.peek().column);
      }
      const nameToken = this.advance();
      className = nameToken.value;

      // Check for instance count with # token
      if (this.check(TokenType.Identifier) && this.peek().value === '#') {
        this.advance(); // consume #
        const countToken = this.consume(TokenType.Number, 'Expected instance count');
        instanceCount = countToken.value;
      }

      this.consume(TokenType.RightParen, 'Expected )');
    }

    const obj = new TonObject(className, instanceCount);

    while (!this.check(TokenType.RightBrace) && !this.isAtEnd()) {
      // Check for child object
      if (this.check(TokenType.LeftBrace)) {
        // Child object
        const childObj = this.parseObject();
        obj.addChild(childObj);

        // Check for comma
        if (this.check(TokenType.Comma)) {
          this.advance();
        }
        continue;
      }

      // Check for optional @ prefix (property marker)
      const hasAtPrefix = this.check(TokenType.AtSign);
      if (hasAtPrefix) {
        this.advance(); // consume @
      }

      // Check for optional / prefix (path marker)
      const hasSlashPrefix = this.check(TokenType.Slash);
      if (hasSlashPrefix) {
        this.advance(); // consume /
      }

      // Parse property name (can be identifier, string, number, guid, or keywords)
      const nameToken = this.advance();
      if (nameToken.type !== TokenType.Identifier &&
          nameToken.type !== TokenType.String &&
          nameToken.type !== TokenType.Number &&
          nameToken.type !== TokenType.Guid &&
          nameToken.type !== TokenType.ClassName &&
          nameToken.type !== TokenType.Boolean &&
          nameToken.type !== TokenType.Null &&
          nameToken.type !== TokenType.Undefined) {
        throw new TonParseError(
          'Expected property name',
          nameToken.line,
          nameToken.column
        );
      }

      // Convert keyword token values back to string names
      let name: string;
      if (nameToken.type === TokenType.Boolean) {
        name = nameToken.value ? 'true' : 'false';
      } else if (nameToken.type === TokenType.Null) {
        name = 'null';
      } else if (nameToken.type === TokenType.Undefined) {
        name = 'undefined';
      } else {
        name = String(nameToken.value);
      }

      // Check for type annotation or separator
      let typeHint: string | undefined;
      let separatorConsumed = false;

      if (this.check(TokenType.Colon)) {
        this.advance(); // consume first :

        // Check if next token is a type identifier (for type annotation)
        if (this.check(TokenType.Identifier)) {
          // This is a type annotation like name:string
          typeHint = this.advance().value;
          // Now expect another : or =
          if (this.check(TokenType.Colon)) {
            this.advance(); // consume second :
            separatorConsumed = true;
          } else if (this.check(TokenType.Equals)) {
            this.advance(); // consume =
            separatorConsumed = true;
          }
        } else {
          // The : was the separator itself (like name: value)
          separatorConsumed = true;
        }
      }

      // If separator not yet consumed, expect equals sign
      if (!separatorConsumed) {
        this.consume(TokenType.Equals, 'Expected Equals (=) or Colon (:) after property name');
      }

      // Parse property value
      const value = this.parseValue();

      // Wrap value in TonValue (like C# implementation does)
      let tonValue: TonValue;
      if (value instanceof TonValue) {
        tonValue = value;
      } else {
        tonValue = TonValue.from(value);
      }

      if (typeHint) {
        tonValue.typeHint = typeHint;
      }

      obj.set(name, tonValue);

      // Check for comma or closing brace
      if (!this.check(TokenType.RightBrace)) {
        if (this.check(TokenType.Comma)) {
          this.advance();
          // Trailing commas are allowed
        } else if (this.check(TokenType.LeftBrace)) {
          // Next token is {, which will be parsed as a child object in the next iteration
          // No comma required before child objects
        } else {
          // Require comma between properties
          const next = this.peek();
          if (next.type !== TokenType.RightBrace) {
            throw new TonParseError(
              'Expected comma or }',
              next.line,
              next.column
            );
          }
        }
      }
    }

    this.consume(TokenType.RightBrace, 'Expected }');
    return obj;
  }

  private parseArray(): TonArray {
    this.consume(TokenType.LeftBracket, 'Expected [');
    const arr = new TonArray();

    while (!this.check(TokenType.RightBracket) && !this.isAtEnd()) {
      arr.add(this.parseValue());

      if (!this.check(TokenType.RightBracket)) {
        if (this.check(TokenType.Comma)) {
          this.advance();
          // Trailing commas are allowed
        } else {
          const next = this.peek();
          if (next.type !== TokenType.RightBracket) {
            throw new TonParseError(
              'Expected comma or ]',
              next.line,
              next.column
            );
          }
        }
      }
    }

    this.consume(TokenType.RightBracket, 'Expected ]');
    return arr;
  }

  private parseEnum(): TonValue {
    const token = this.advance();
    return new TonValue(token.value, 'enum');
  }

  private parseHintedValue(): TonValue {
    const hintToken = this.advance();
    let typeHint: string;

    switch (hintToken.type) {
      case TokenType.StringHint:
        typeHint = 'string';
        break;
      case TokenType.NumberHint:
        typeHint = 'number';
        break;
      case TokenType.BooleanHint:
        typeHint = 'boolean';
        break;
      case TokenType.DateHint:
        typeHint = 'date';
        break;
      default:
        typeHint = 'unknown';
    }

    const value = this.parseValue();
    if (value instanceof TonValue) {
      value.typeHint = typeHint;
    }

    return value;
  }

  private parseTypedObject(): TonObject {
    // Handle both (ClassName) and ClassName syntaxes
    let className: string;
    let instanceCount: number | undefined;

    if (this.check(TokenType.LeftParen)) {
      // Format: (ClassName) or (ClassName#count)
      this.advance(); // consume (

      // Class name can be either ClassName or Identifier token type
      if (!this.check(TokenType.ClassName) && !this.check(TokenType.Identifier)) {
        throw new TonParseError('Expected class name', this.peek().line, this.peek().column);
      }
      const nameToken = this.advance();
      className = nameToken.value;

      // Check for instance count with # token
      if (this.check(TokenType.Identifier) && this.peek().value === '#') {
        this.advance(); // consume #
        const countToken = this.consume(TokenType.Number, 'Expected instance count');
        instanceCount = countToken.value;
      }

      this.consume(TokenType.RightParen, 'Expected )');
    } else {
      // Format: ClassName (legacy support)
      const nameToken = this.advance();
      className = nameToken.value;

      // Check for instance count with parentheses
      if (this.check(TokenType.LeftParen)) {
        this.advance(); // consume (
        const countToken = this.consume(TokenType.Number, 'Expected instance count');
        instanceCount = countToken.value;
        this.consume(TokenType.RightParen, 'Expected )');
      }
    }

    const obj = this.parseObject();
    obj.className = className;
    obj.instanceCount = instanceCount;

    return obj;
  }

  private consume(type: TokenType, message: string): Token {
    if (this.check(type)) {
      return this.advance();
    }

    const token = this.peek();
    throw new TonParseError(message, token.line, token.column);
  }

  private check(type: TokenType): boolean {
    if (this.isAtEnd()) return false;
    return this.peek().type === type;
  }

  private advance(): Token {
    if (!this.isAtEnd()) this.current++;
    return this.previous();
  }

  private peek(): Token {
    return this.tokens[this.current];
  }

  private previous(): Token {
    return this.tokens[this.current - 1];
  }

  private isAtEnd(): boolean {
    return this.peek().type === TokenType.EndOfFile;
  }
}